File: Practice/36/C++/main.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <cmath>
#include <fstream>

const auto PI = 3.141592653589793;

enum class coordinates_system { Cartesian, Polar };

class Point {
public:
    double x, y;
    Point(double = 0, double = 0, coordinates_system = coordinates_system::Cartesian);
    bool operator==(const Point&) const;
    bool operator!=(const Point&) const;

    double get_x() const {
        return x;
    };

    double get_y() const {
        return y;
    };

    double get_r() const {
        return hypot(x, y);
    };

    double get_phi() const {
        return atan2(y, x);
    };

    void set_x(double x_new) {
        x = x_new;
    };

    void set_y(double y_new) {
        y = y_new;
    };

    void set_r(double);
    void set_phi(double);

private:
    std::string chunga{"changa - easter egg"};
};

Point::Point(double a1, double a2, coordinates_system coord_system) {
    if (coord_system == coordinates_system::Cartesian) {
        x = a1;
        y = a2;
    }

    else {
        x = cos(a2) * a1;
        y = sin(a2) * a1;
    }
}

void Point::set_r(double r) {
    double phi = get_phi();

    x = cos(phi) * r;
    y = sin(phi) * r;
}

void Point::set_phi(double phi) {
    double r = get_r();

    x = cos(phi) * r;
    y = sin(phi) * r;
}

bool operator==(const Point& lhs, const Point& rhs) {
    return (fabs(lhs.x - rhs.x) < 1e-10) && (fabs(lhs.y - rhs.y) < 1e-10);
}

bool operator!=(const Point& lhs, const Point& rhs) {
    return !((fabs(lhs.x - rhs.x) < 1e-10) && (fabs(lhs.y - rhs.y) < 1e-10));
}

std::ostream& operator<<(std::ostream& output, const Point& point) {
    output << '(' << point.get_x() << ',' << point.get_y() << ')';
    return output;
}

std::istream& operator>>(std::istream& input, Point& point) {
    double buf;
    input.ignore(1);
    input >> buf;
    point.set_x(buf);
    input.ignore(1);
    input >> buf;
    point.set_y(buf);
    input.ignore(1);
    return input;
}


int main() {
    std::vector<Point> original;
    std::ifstream fin("data.txt");
    if (!fin.is_open()) {
        std::cout << "Can't open file" << std::endl;
        return 1;
    } else {
        while (!fin.eof()) {
            Point p;
            fin >> p;
            fin.ignore(2);
            original.push_back(p);
        }
        fin.close();
    }

    std::vector<Point> simulacrum(original);
    for (auto& p : simulacrum) {
        std::cout << p;
        p.set_x(p.get_x() + 10);
        p.set_phi(p.get_phi() + 180*PI/180);
        p.set_y(-p.get_y());
        p.set_x(-p.get_x() - 10);
        std::cout << p << std::endl;
    }

    if (std::equal(original.begin(), original.end(), simulacrum.begin()))
        std::cout << "\nIt works!\n";
    else
        std::cout << "\nIt not works!\n";
}

File: Practice/36/python/main.py
# This Python file uses the following encoding: utf-8
import math
import copy


cordsys = ['Cartesian', 'Polar']

class Point:
    def __init__(self, a1 = 0, a2 = 0, coordinates_system = 'Cartesian'):
        if type(a1) == str:
            a1.replace(',', ' ')
            a1.replace('(', '')
            a1.replace(')', '')
            self.x = float( a1.split()[0] )
            self.y = float( a1.split()[1] )
        else:
            if (coordinates_system == 'Cartesian'):
                self.x = a1
                self.y = a2
            else:
                self.x = math.cos(a2) * a1
                self.y = math.sin(a2) * a1

    def __eq__(self, other):
        return (abs(self.x - other.x) < 1e-10) and (abs(self.y - other.y) < 1e-10)

    def __ne__(self, other):
        return not(self == other)

    def get_x(self):
        return self.x

    def get_y(self):
        return self.y

    def get_r(self):
        return math.hypot(self.x, self.y)

    def get_phi(self):
        return math.atan2(self.y, self.x)

    def set_x(self, x):
        self.x = x

    def set_y(self, y):
        self.y = y

    def set_r(self, r):
        phi = self.get_phi()

        self.x = math.cos(phi) * r
        self.y = math.sin(phi) * r

    def set_phi(self, phi):
        r = self.get_r()

        self.x = math.cos(phi) * r
        self.y = math.sin(phi) * r

    def __repr__(self):
        return 'Point({1},{2})'.format(self.x, self.y)

    def __str__(self):
        return '({1},{2})'.format(self.x, self.y)

with open('data.txt') as fin:
    original = [Point(p) for p in fin.readline().split(', ')]

simulacrum = copy.deepcopy(original)
for p in simulacrum:
    print(p, end='')
    p.set_x(p.get_x() + 10)
    p.set_phi(p.get_phi() + 180*math.pi/180)
    p.set_y(-p.get_y())
    p.set_x(-p.get_x() - 10)
    print(p)

print('\nIt works!\n' if simulacrum == original else '\nIt not works!\n')
